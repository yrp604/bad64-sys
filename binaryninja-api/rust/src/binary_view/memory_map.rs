use crate::binary_view::BinaryView;
use crate::data_buffer::DataBuffer;
use crate::file_accessor::{Accessor, FileAccessor};
use crate::rc::Ref;
use crate::segment::SegmentFlags;
use crate::string::{BnString, IntoCStr};
use binaryninjacore_sys::*;

#[derive(PartialEq, Eq, Hash)]
pub struct MemoryMap {
    view: Ref<BinaryView>,
}

impl MemoryMap {
    pub fn new(view: Ref<BinaryView>) -> Self {
        Self { view }
    }

    // TODO: There does not seem to be a way to enumerate memory regions.

    /// JSON string representation of the base [`MemoryMap`], consisting of unresolved auto and user segments.
    pub fn base_description(&self) -> String {
        let desc_raw = unsafe { BNGetBaseMemoryMapDescription(self.view.handle) };
        unsafe { BnString::into_string(desc_raw) }
    }

    /// JSON string representation of the [`MemoryMap`].
    pub fn description(&self) -> String {
        let desc_raw = unsafe { BNGetMemoryMapDescription(self.view.handle) };
        unsafe { BnString::into_string(desc_raw) }
    }

    // When enabled, the memory map will present a simplified, logical view that merges and abstracts virtual memory
    // regions based on criteria such as contiguity and flag consistency. This view is designed to provide a higher-level
    // representation for user analysis, hiding underlying mapping details.
    //
    // When disabled, the memory map will revert to displaying the virtual view, which corresponds directly to the individual
    // segments mapped from the raw file without any merging or abstraction.
    pub fn set_logical_enabled(&mut self, enabled: bool) {
        unsafe { BNSetLogicalMemoryMapEnabled(self.view.handle, enabled) };
    }

    /// Whether the memory map is activated for the associated view.
    pub fn is_activated(&self) -> bool {
        unsafe { BNIsMemoryMapActivated(self.view.handle) }
    }

    pub fn add_binary_memory_region(
        &mut self,
        name: &str,
        start: u64,
        view: &BinaryView,
        segment_flags: Option<SegmentFlags>,
    ) -> bool {
        let name_raw = name.to_cstr();
        unsafe {
            BNAddBinaryMemoryRegion(
                self.view.handle,
                name_raw.as_ptr(),
                start,
                view.handle,
                segment_flags.unwrap_or_default().into_raw(),
            )
        }
    }

    /// Adds the memory region using a [`DataBuffer`].
    ///
    /// This will add the contents of the [`DataBuffer`] to the database.
    pub fn add_data_memory_region(
        &mut self,
        name: &str,
        start: u64,
        data: &DataBuffer,
        segment_flags: Option<SegmentFlags>,
    ) -> bool {
        let name_raw = name.to_cstr();
        unsafe {
            BNAddDataMemoryRegion(
                self.view.handle,
                name_raw.as_ptr(),
                start,
                data.as_raw(),
                segment_flags.unwrap_or_default().into_raw(),
            )
        }
    }

    // TODO: This really cant be safe until BNFileAccessor is ARC'd and can be freed. Probably need another thing
    // TODO: Ontop of a file accessor in the core that would manage it. (I.e. BNFileAccessorHandle) or something.
    /// Adds the memory region using a [`FileAccessor`].
    ///
    /// This does not add the region contents to the database, instead accesses to the contents
    /// are done "remotely" to a [`FileAccessor`].
    ///
    /// NOTE: The [`FileAccessor`] MUST live as long as the region is available, currently there is no gurentee by
    /// the type checker that the file accessor is tied to that of the memory region.
    pub fn add_remote_memory_region<A: Accessor>(
        &mut self,
        name: &str,
        start: u64,
        accessor: &mut FileAccessor<A>,
        segment_flags: Option<SegmentFlags>,
    ) -> bool {
        let name_raw = name.to_cstr();
        unsafe {
            BNAddRemoteMemoryRegion(
                self.view.handle,
                name_raw.as_ptr(),
                start,
                &mut accessor.raw,
                segment_flags.unwrap_or_default().into_raw(),
            )
        }
    }

    /// Adds an unbacked memory region with a given length and fill byte.
    pub fn add_unbacked_memory_region(
        &mut self,
        name: &str,
        start: u64,
        length: u64,
        segment_flags: Option<SegmentFlags>,
        fill: Option<u8>,
    ) -> bool {
        let name_raw = name.to_cstr();
        unsafe {
            BNAddUnbackedMemoryRegion(
                self.view.handle,
                name_raw.as_ptr(),
                start,
                length,
                segment_flags.unwrap_or_default().into_raw(),
                fill.unwrap_or_default(),
            )
        }
    }

    pub fn remove_memory_region(&mut self, name: &str) -> bool {
        let name_raw = name.to_cstr();
        unsafe { BNRemoveMemoryRegion(self.view.handle, name_raw.as_ptr()) }
    }

    pub fn active_memory_region_at(&self, addr: u64) -> String {
        unsafe {
            let name_raw = BNGetActiveMemoryRegionAt(self.view.handle, addr);
            BnString::into_string(name_raw)
        }
    }

    pub fn memory_region_flags(&self, name: &str) -> SegmentFlags {
        let name_raw = name.to_cstr();
        let flags_raw = unsafe { BNGetMemoryRegionFlags(self.view.handle, name_raw.as_ptr()) };
        SegmentFlags::from_raw(flags_raw)
    }

    pub fn set_memory_region_flags(&mut self, name: &str, flags: SegmentFlags) -> bool {
        let name_raw = name.to_cstr();
        unsafe { BNSetMemoryRegionFlags(self.view.handle, name_raw.as_ptr(), flags.into_raw()) }
    }

    pub fn is_memory_region_enabled(&self, name: &str) -> bool {
        let name_raw = name.to_cstr();
        unsafe { BNIsMemoryRegionEnabled(self.view.handle, name_raw.as_ptr()) }
    }

    pub fn set_memory_region_enabled(&mut self, name: &str, enabled: bool) -> bool {
        let name_raw = name.to_cstr();
        unsafe { BNSetMemoryRegionEnabled(self.view.handle, name_raw.as_ptr(), enabled) }
    }

    // TODO: Should we just call this is_memory_region_relocatable?
    pub fn is_memory_region_rebaseable(&self, name: &str) -> bool {
        let name_raw = name.to_cstr();
        unsafe { BNIsMemoryRegionRebaseable(self.view.handle, name_raw.as_ptr()) }
    }

    pub fn set_memory_region_rebaseable(&mut self, name: &str, enabled: bool) -> bool {
        let name_raw = name.to_cstr();
        unsafe { BNSetMemoryRegionRebaseable(self.view.handle, name_raw.as_ptr(), enabled) }
    }

    pub fn memory_region_fill(&self, name: &str) -> u8 {
        let name_raw = name.to_cstr();
        unsafe { BNGetMemoryRegionFill(self.view.handle, name_raw.as_ptr()) }
    }

    pub fn set_memory_region_fill(&mut self, name: &str, fill: u8) -> bool {
        let name_raw = name.to_cstr();
        unsafe { BNSetMemoryRegionFill(self.view.handle, name_raw.as_ptr(), fill) }
    }

    pub fn reset(&mut self) {
        unsafe { BNResetMemoryMap(self.view.handle) }
    }
}
