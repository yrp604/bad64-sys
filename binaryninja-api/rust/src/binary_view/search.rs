use serde_derive::{Deserialize, Serialize};

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SearchQuery {
    /// ex. "42 2e 64 65 ?? 75 67 24"
    pattern: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    start: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    end: Option<u64>,
    #[serde(rename = "ignoreCase")]
    ignore_case: bool,
    raw: bool,
    overlap: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    align: Option<u64>,
}

impl SearchQuery {
    pub fn new(pattern: impl Into<String>) -> Self {
        Self {
            pattern: pattern.into(),
            ..Default::default()
        }
    }

    /// Set the starting address for the search
    pub fn start(mut self, addr: u64) -> Self {
        self.start = Some(addr);
        self
    }

    /// Set the ending address for the search (inclusive)
    pub fn end(mut self, addr: u64) -> Self {
        self.end = Some(addr);
        self
    }

    /// Set whether to interpret the pattern as a raw string
    pub fn raw(mut self, raw: bool) -> Self {
        self.raw = raw;
        self
    }

    /// Set whether to perform case-insensitive matching
    pub fn ignore_case(mut self, ignore_case: bool) -> Self {
        self.ignore_case = ignore_case;
        self
    }

    /// Set whether to allow matches to overlap
    pub fn overlap(mut self, overlap: bool) -> Self {
        self.overlap = overlap;
        self
    }

    /// Set the alignment of matches (must be a power of 2)
    pub fn align(mut self, align: u64) -> Self {
        // Validate that align is a power of 2
        if align != 0 && (align & (align - 1)) == 0 {
            self.align = Some(align);
        }
        self
    }
}

impl SearchQuery {
    /// Serialize the query to a JSON string
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).expect("failed to serialize search query")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_search_query_builder() {
        let query = SearchQuery::new("test pattern")
            .start(0x1000)
            .end(0x2000)
            .raw(true)
            .ignore_case(true)
            .overlap(false)
            .align(16);

        assert_eq!(query.pattern, "test pattern");
        assert_eq!(query.start, Some(0x1000));
        assert_eq!(query.end, Some(0x2000));
        assert!(query.raw);
        assert!(query.ignore_case);
        assert!(!query.overlap);
        assert_eq!(query.align, Some(16));
    }

    #[test]
    fn test_search_query_json() {
        let query = SearchQuery::new("test").start(0x1000).align(8);

        let json = query.to_json();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();

        assert_eq!(parsed["pattern"], "test");
        assert_eq!(parsed["start"], 4096);
        assert_eq!(parsed["align"], 8);
        assert!(!parsed.as_object().unwrap().contains_key("end"));
    }
}
