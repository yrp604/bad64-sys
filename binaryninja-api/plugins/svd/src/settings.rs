use binaryninja::binary_view::BinaryView;
use binaryninja::settings::{QueryOptions, Settings};
use serde_json::json;
use std::path::PathBuf;

#[derive(Debug, Clone)]
pub struct LoadSettings {
    pub add_backing_regions: bool,
    pub add_bitfields: bool,
    pub add_comments: bool,
    pub auto_load_file: Option<PathBuf>,
}

impl LoadSettings {
    pub const ADD_BACKING_REGIONS_DEFAULT: bool = true;
    pub const ADD_BACKING_REGIONS_SETTING: &'static str = "analysis.svd.addBackingRegions";
    pub const ADD_BITFIELDS_DEFAULT: bool = true;
    pub const ADD_BITFIELDS_SETTING: &'static str = "analysis.svd.addBitfields";
    pub const ADD_COMMENTS_DEFAULT: bool = true;
    pub const ADD_COMMENTS_SETTING: &'static str = "analysis.svd.addComments";
    pub const AUTO_LOAD_FILE_DEFAULT: &'static str = "";
    pub const AUTO_LOAD_FILE_SETTING: &'static str = "analysis.svd.autoLoadFile";

    pub fn register() {
        let bn_settings = Settings::new();

        let add_backing_region_props = json!({
            "title" : "Add Backing Regions",
            "type" : "boolean",
            "default" : Self::ADD_BACKING_REGIONS_DEFAULT,
            "description" : "Whether to add backing regions. Backing regions allow you to write to the underlying memory of a view, but will take up space in the BNDB.",
        });
        bn_settings.register_setting_json(
            Self::ADD_BACKING_REGIONS_SETTING,
            &add_backing_region_props.to_string(),
        );

        let add_bitfields_props = json!({
            "title" : "Add Bitfields",
            "type" : "boolean",
            "default" : Self::ADD_BITFIELDS_DEFAULT,
            "description" : "Whether to add bitfields. Bitfields are not supported by Binary Ninja, so this is a workaround using unions.",
        });
        bn_settings.register_setting_json(
            Self::ADD_BITFIELDS_SETTING,
            &add_bitfields_props.to_string(),
        );

        let add_comments_props = json!({
            "title" : "Add Comments",
            "type" : "boolean",
            "default" : Self::ADD_COMMENTS_DEFAULT,
            "description" : "Whether to add comments. If you see comment placement is off, try disabling this.",
        });
        bn_settings
            .register_setting_json(Self::ADD_COMMENTS_SETTING, &add_comments_props.to_string());

        let file_props = json!({
            "title" : "SVD File",
            "type" : "string",
            "default" : Self::AUTO_LOAD_FILE_DEFAULT,
            "description" : "The SVD File to automatically load when opening the view.",
            "uiSelectionAction" : "file"
        });
        bn_settings.register_setting_json(Self::AUTO_LOAD_FILE_SETTING, &file_props.to_string());
    }

    pub fn from_view_settings(view: &BinaryView) -> Self {
        let mut load_settings = LoadSettings::default();
        let settings = Settings::new();
        let mut query_opts = QueryOptions::new_with_view(view);
        if settings.contains(Self::ADD_BACKING_REGIONS_SETTING) {
            load_settings.add_backing_regions =
                settings.get_bool_with_opts(Self::ADD_BACKING_REGIONS_SETTING, &mut query_opts);
        }
        if settings.contains(Self::ADD_BITFIELDS_SETTING) {
            load_settings.add_bitfields =
                settings.get_bool_with_opts(Self::ADD_BITFIELDS_SETTING, &mut query_opts);
        }
        if settings.contains(Self::ADD_COMMENTS_SETTING) {
            load_settings.add_comments =
                settings.get_bool_with_opts(Self::ADD_COMMENTS_SETTING, &mut query_opts);
        }
        if settings.contains(Self::AUTO_LOAD_FILE_SETTING) {
            let path_str =
                settings.get_string_with_opts(Self::AUTO_LOAD_FILE_SETTING, &mut query_opts);
            if !path_str.is_empty() {
                let path = PathBuf::from(path_str.to_string());
                load_settings.auto_load_file = Some(path);
            }
        }
        load_settings
    }
}

impl Default for LoadSettings {
    fn default() -> Self {
        Self {
            add_backing_regions: Self::ADD_BACKING_REGIONS_DEFAULT,
            add_bitfields: Self::ADD_BITFIELDS_DEFAULT,
            add_comments: Self::ADD_COMMENTS_DEFAULT,
            auto_load_file: None,
        }
    }
}
