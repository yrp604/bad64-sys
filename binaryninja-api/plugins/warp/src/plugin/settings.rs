use binaryninja::settings::{QueryOptions, Settings as BNSettings};
use serde_json::json;
use std::string::ToString;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct PluginSettings {
    /// Whether to load bundled WARP files on startup. Turn this off if you want to manually load them.
    ///
    /// This is set to [PluginSettings::LOAD_BUNDLED_FILES_DEFAULT] by default.
    pub load_bundled_files: bool,
    /// Whether to load user WARP files on startup. Turn this off if you want to manually load them.
    ///
    /// This is set to [PluginSettings::LOAD_USER_FILES_DEFAULT] by default.
    pub load_user_files: bool,
    /// The WARP server to use.
    ///
    /// This is set to [PluginSettings::SERVER_URL_DEFAULT] by default.
    pub server_url: String,
    /// The API key to use for the selected WARP server, if not specified, you will be unable to push data and may be rate-limited.
    ///
    /// This is set to [PluginSettings::SERVER_API_KEY_DEFAULT] by default.
    pub server_api_key: Option<String>,
    pub second_server_url: Option<String>,
    pub second_server_api_key: Option<String>,
    /// A source must have at least one of these tags to be considered a valid source.
    ///
    /// This is set to [PluginSettings::SOURCE_TAGS_DEFAULT] by default.
    pub whitelisted_source_tags: Vec<String>,
    /// Whether to allow networked WARP requests. Turning this off will not disable local WARP functionality.
    ///
    /// This is set to [PluginSettings::ENABLE_SERVER_DEFAULT] by default.
    pub enable_server: bool,
}

impl PluginSettings {
    pub const WHITELISTED_SOURCE_TAGS_DEFAULT: Vec<String> = vec![];
    pub const WHITELISTED_SOURCE_TAGS_SETTING: &'static str = "analysis.warp.whitelistedSourceTags";
    pub const LOAD_BUNDLED_FILES_DEFAULT: bool = true;
    pub const LOAD_BUNDLED_FILES_SETTING: &'static str = "analysis.warp.loadBundledFiles";
    pub const LOAD_USER_FILES_DEFAULT: bool = true;
    pub const LOAD_USER_FILES_SETTING: &'static str = "analysis.warp.loadUserFiles";
    pub const SERVER_URL_DEFAULT: &'static str = "https://warp.binary.ninja";
    pub const SERVER_URL_SETTING: &'static str = "analysis.warp.serverUrl";
    pub const SERVER_API_KEY_DEFAULT: Option<String> = None;
    pub const SERVER_API_KEY_SETTING: &'static str = "analysis.warp.serverApiKey";
    pub const SECONDARY_SERVER_URL_DEFAULT: Option<String> = None;
    pub const SECONDARY_SERVER_URL_SETTING: &'static str = "analysis.warp.secondServerUrl";
    pub const SECONDARY_SERVER_API_KEY_DEFAULT: Option<String> = None;
    pub const SECONDARY_SERVER_API_KEY_SETTING: &'static str = "analysis.warp.secondServerApiKey";
    pub const ENABLE_SERVER_DEFAULT: bool = false;
    pub const ENABLE_SERVER_SETTING: &'static str = "network.enableWARP";

    pub fn register(bn_settings: &mut BNSettings) {
        let whitelisted_source_tags_prop = json!({
            "title" : "Blacklisted Sources",
            "type" : "array",
            "default" : Self::WHITELISTED_SOURCE_TAGS_DEFAULT,
            "description" : "Add a sources UUID to this list to blacklist it from being considered a valid source. This is useful for sources that are known to be false positives.",
            "ignore" : [],
        });
        bn_settings.register_setting_json(
            Self::WHITELISTED_SOURCE_TAGS_SETTING,
            &whitelisted_source_tags_prop.to_string(),
        );
        let load_bundled_files_prop = json!({
            "title" : "Load Bundled Files",
            "type" : "boolean",
            "default" : Self::LOAD_BUNDLED_FILES_DEFAULT,
            "description" : "Whether to load bundled WARP files on startup. Turn this off if you want to manually load them.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::LOAD_BUNDLED_FILES_SETTING,
            &load_bundled_files_prop.to_string(),
        );
        let load_user_files_prop = json!({
            "title" : "Load User Files",
            "type" : "boolean",
            "default" : Self::LOAD_USER_FILES_DEFAULT,
            "description" : "Whether to load user WARP files on startup. Turn this off if you want to manually load them.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::LOAD_USER_FILES_SETTING,
            &load_user_files_prop.to_string(),
        );
        let server_url_prop = json!({
            "title" : "Server URL",
            "type" : "string",
            "default" : Self::SERVER_URL_DEFAULT,
            "description" : "The WARP server to use.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(Self::SERVER_URL_SETTING, &server_url_prop.to_string());
        let server_api_key_prop = json!({
            "title" : "Server API Key",
            "type" : "string",
            "default" : Self::SERVER_API_KEY_DEFAULT,
            "description" : "The API key to use for the selected WARP server, if not specified you will be unable to push data, and may be rate limited.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "hidden": true,
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::SERVER_API_KEY_SETTING,
            &server_api_key_prop.to_string(),
        );
        let second_server_url_prop = json!({
            "title" : "Secondary Server URL",
            "type" : "string",
            "default" : Self::SECONDARY_SERVER_URL_DEFAULT,
            "description" : "",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::SECONDARY_SERVER_URL_SETTING,
            &second_server_url_prop.to_string(),
        );
        let second_server_api_key_prop = json!({
            "title" : "Secondary Server API Key",
            "type" : "string",
            "default" : Self::SECONDARY_SERVER_API_KEY_DEFAULT,
            "description" : "",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "hidden": true,
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::SECONDARY_SERVER_API_KEY_SETTING,
            &second_server_api_key_prop.to_string(),
        );
        let server_enabled_prop = json!({
            "title" : "Enable WARP",
            "type" : "boolean",
            "default" : Self::ENABLE_SERVER_DEFAULT,
            "description" : "Whether or not to allow networked WARP requests. Turning this off will not disable local WARP functionality.",
            "ignore" : ["SettingsProjectScope", "SettingsResourceScope"],
            "requiresRestart" : true
        });
        bn_settings.register_setting_json(
            Self::ENABLE_SERVER_SETTING,
            &server_enabled_prop.to_string(),
        );
    }

    /// Retrieve plugin settings from [`BNSettings`].
    pub fn from_settings(bn_settings: &BNSettings, query_opts: &mut QueryOptions) -> Self {
        let mut settings = PluginSettings::default();
        if bn_settings.contains(Self::LOAD_BUNDLED_FILES_SETTING) {
            settings.load_bundled_files = bn_settings.get_bool(Self::LOAD_BUNDLED_FILES_SETTING);
        }
        if bn_settings.contains(Self::LOAD_USER_FILES_SETTING) {
            settings.load_user_files = bn_settings.get_bool(Self::LOAD_USER_FILES_SETTING);
        }
        if bn_settings.contains(Self::SERVER_URL_SETTING) {
            settings.server_url = bn_settings.get_string(Self::SERVER_URL_SETTING);
        }
        if bn_settings.contains(Self::SERVER_API_KEY_SETTING) {
            let server_api_key_str = bn_settings.get_string(Self::SERVER_API_KEY_SETTING);
            if !server_api_key_str.is_empty() {
                settings.server_api_key = Some(server_api_key_str);
            }
        }
        if bn_settings.contains(Self::SECONDARY_SERVER_URL_SETTING) {
            let server_api_key_str = bn_settings.get_string(Self::SECONDARY_SERVER_URL_SETTING);
            if !server_api_key_str.is_empty() {
                settings.second_server_url = Some(server_api_key_str);
            }
        }
        if bn_settings.contains(Self::SECONDARY_SERVER_API_KEY_SETTING) {
            let server_api_key_str = bn_settings.get_string(Self::SECONDARY_SERVER_API_KEY_SETTING);
            if !server_api_key_str.is_empty() {
                settings.second_server_api_key = Some(server_api_key_str);
            }
        }
        if bn_settings.contains(Self::ENABLE_SERVER_SETTING) {
            settings.enable_server = bn_settings.get_bool(Self::ENABLE_SERVER_SETTING);
        }

        if bn_settings.contains(Self::WHITELISTED_SOURCE_TAGS_SETTING) {
            let whitelisted_source_tags_str = bn_settings
                .get_string_list_with_opts(Self::WHITELISTED_SOURCE_TAGS_SETTING, query_opts);
            settings.whitelisted_source_tags = whitelisted_source_tags_str
                .iter()
                .map(|s| s.to_string())
                .collect();
        }
        settings
    }
}

impl Default for PluginSettings {
    fn default() -> Self {
        Self {
            whitelisted_source_tags: PluginSettings::WHITELISTED_SOURCE_TAGS_DEFAULT,
            load_bundled_files: PluginSettings::LOAD_BUNDLED_FILES_DEFAULT,
            load_user_files: PluginSettings::LOAD_USER_FILES_DEFAULT,
            server_url: PluginSettings::SERVER_URL_DEFAULT.to_string(),
            server_api_key: PluginSettings::SERVER_API_KEY_DEFAULT,
            second_server_url: PluginSettings::SECONDARY_SERVER_URL_DEFAULT,
            second_server_api_key: PluginSettings::SECONDARY_SERVER_API_KEY_DEFAULT,
            enable_server: PluginSettings::ENABLE_SERVER_DEFAULT,
        }
    }
}
