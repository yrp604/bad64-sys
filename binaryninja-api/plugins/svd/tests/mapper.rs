use binaryninja::binary_view::BinaryViewExt;
use binaryninja::binary_view::{BinaryView, BinaryViewBase};
use binaryninja::file_metadata::FileMetadata;
use binaryninja::headless::Session;
use std::path::PathBuf;
use svd_ninja::mapper::DeviceMapper;
use svd_ninja::settings::LoadSettings;

// These are the target files present in OUT_DIR
// Add the files to fixtures
static TARGET_FILES: [&str; 2] = ["ARM_Sample.svd", "esp32c2.svd"];

#[test]
fn insta_snapshots() {
    let _session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let svd_str = std::fs::read_to_string(&path).expect("Failed to read svd file");
        let device = svd_parser::parse(&svd_str).expect("Failed to parse svd file");
        let view = BinaryView::from_data(&FileMetadata::new(), &[]).expect("Failed to create view");
        let address_size = view.address_size();
        DeviceMapper::new(LoadSettings::default(), address_size, device).map_to_view(&view);

        let types = view
            .types()
            .to_vec()
            .iter()
            .map(|ty| ty.ty.to_string())
            .collect::<Vec<String>>();
        let types_snapshot_name = format!("{}_types", file_name);
        insta::assert_debug_snapshot!(types_snapshot_name, types);

        // TODO: This does not work yet because memory map is not turned on for raw view?
        // let value: Value = serde_json::from_str(&view.memory_map().description()).unwrap();
        // let map = view.memory_map().description();
        // let memory_snapshot_name = format!("{}_memory_map", file_name);
        // insta::assert_debug_snapshot!(memory_snapshot_name, value);
    }
}

#[test]
fn test_bitfield_unions() {
    let _session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    let path = out_dir.join("ARM_Sample.svd");
    let svd_str = std::fs::read_to_string(&path).expect("Failed to read svd file");
    let device = svd_parser::parse(&svd_str).expect("Failed to parse svd file");
    let view = BinaryView::from_data(&FileMetadata::new(), &[]).expect("Failed to create view");
    let address_size = view.address_size();
    let mapper = DeviceMapper::new(LoadSettings::default(), address_size, device.clone());

    let peripheral = device.get_peripheral("TIMER0").unwrap();
    let register = peripheral.get_register("SR").unwrap();
    let register_ty = mapper.register_type(&register);
    insta::assert_debug_snapshot!("bitfield_union_types", register_ty);
}
