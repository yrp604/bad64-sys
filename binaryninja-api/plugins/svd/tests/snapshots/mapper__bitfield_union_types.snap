---
source: plugins/svd/tests/mapper.rs
expression: register_ty
---
Type {
    type_class: StructureTypeClass,
    width: 2,
    alignment: 1,
    is_signed: true (255 confidence),
    is_const: false (255 confidence),
    is_volatile: false (255 confidence),
    child_type: None,
    calling_convention: None,
    parameters: Some(
        [],
    ),
    has_variable_arguments: false (255 confidence),
    can_return: true (255 confidence),
    pure: false (255 confidence),
    get_structure: Some(
        Structure {
            width: 2,
            structure_type: StructStructureType,
            base_structures: [],
            members: [
                StructureMember {
                    ty: Type { type_class: StructureTypeClass, width: 1, alignment: 1, is_signed: true (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: Some(Structure { width: 1, structure_type: UnionStructureType, base_structures: [], members: [StructureMember { ty: Type { type_class: EnumerationTypeClass, width: 1, alignment: 1, is_signed: false (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: None, get_enumeration: Some(Enumeration { members: [EnumerationMember { name: "Stopped", value: 0, default: false }, EnumerationMember { name: "Running", value: 1, default: false }] }), get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence), name: "RUN", offset: 0, access: PublicAccess, scope: NoScope }] }), get_enumeration: None, get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence),
                    name: "bitfield_0x0",
                    offset: 0,
                    access: PublicAccess,
                    scope: NoScope,
                },
                StructureMember {
                    ty: Type { type_class: StructureTypeClass, width: 1, alignment: 1, is_signed: true (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: Some(Structure { width: 1, structure_type: UnionStructureType, base_structures: [], members: [StructureMember { ty: Type { type_class: EnumerationTypeClass, width: 1, alignment: 1, is_signed: false (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: None, get_enumeration: Some(Enumeration { members: [EnumerationMember { name: "No_Match", value: 0, default: false }, EnumerationMember { name: "Match_Hit", value: 1, default: false }] }), get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence), name: "MATCH", offset: 0, access: PublicAccess, scope: NoScope }, StructureMember { ty: Type { type_class: EnumerationTypeClass, width: 1, alignment: 1, is_signed: false (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: None, get_enumeration: Some(Enumeration { members: [EnumerationMember { name: "No_Underflow", value: 0, default: false }, EnumerationMember { name: "Underflow", value: 1, default: false }] }), get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence), name: "UN", offset: 0, access: PublicAccess, scope: NoScope }, StructureMember { ty: Type { type_class: EnumerationTypeClass, width: 1, alignment: 1, is_signed: false (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: None, get_enumeration: Some(Enumeration { members: [EnumerationMember { name: "No_Overflow", value: 0, default: false }, EnumerationMember { name: "Overflow_occured", value: 1, default: false }] }), get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence), name: "OV", offset: 0, access: PublicAccess, scope: NoScope }, StructureMember { ty: Type { type_class: EnumerationTypeClass, width: 1, alignment: 1, is_signed: false (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: None, get_enumeration: Some(Enumeration { members: [EnumerationMember { name: "Ready", value: 0, default: false }, EnumerationMember { name: "In_Reset", value: 1, default: false }] }), get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence), name: "RST", offset: 0, access: PublicAccess, scope: NoScope }, StructureMember { ty: Type { type_class: EnumerationTypeClass, width: 1, alignment: 1, is_signed: false (255 confidence), is_const: false (255 confidence), is_volatile: false (255 confidence), child_type: None, calling_convention: None, parameters: Some([]), has_variable_arguments: false (255 confidence), can_return: true (255 confidence), pure: false (255 confidence), get_structure: None, get_enumeration: Some(Enumeration { members: [EnumerationMember { name: "RELOAD0", value: 0, default: false }, EnumerationMember { name: "RELOAD1", value: 1, default: false }, EnumerationMember { name: "RELOAD2", value: 2, default: false }, EnumerationMember { name: "RELOAD3", value: 3, default: false }] }), get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence), name: "RELOAD", offset: 0, access: PublicAccess, scope: NoScope }] }), get_enumeration: None, get_named_type_reference: None, count: 0, offset: 0, stack_adjustment: 0 (0 confidence), registered_name: None } (255 confidence),
                    name: "bitfield_0x1",
                    offset: 1,
                    access: PublicAccess,
                    scope: NoScope,
                },
            ],
        },
    ),
    get_enumeration: None,
    get_named_type_reference: None,
    count: 0,
    offset: 0,
    stack_adjustment: 0 (0 confidence),
    registered_name: None,
}
